### PR TITLE
feat: add metrics for FallbackConfiguration

### LINKED ISSUE #6082: Expose fallback configuration metrics
We should make it possible to monitor fallback configuration process using Prometheus metrics:

- How many times fallback configuration was built
- How many successful/failed fallback config push attempts we had
- Last time of successful fallback configuration push

### PR DESCRIPTION
**Pull request by czeslavo**
**What this PR does / why we need it**:

Adds new Prometheus metrics covering `FallbackConfiguration` feature:

- `ingress_controller_fallback_translation_count`
- `ingress_controller_fallback_translation_broken_resource_count`
- `ingress_controller_fallback_configuration_push_count`
- `ingress_controller_fallback_configuration_push_last`
- `ingress_controller_fallback_configuration_push_duration_milliseconds`
- `ingress_controller_fallback_configuration_push_broken_resource_count`
- `ingress_controller_fallback_cache_generating_duration_milliseconds`
- `ingress_controller_processed_config_snapshot_cache_hit`
- `ingress_controller_processed_config_snapshot_cache_miss`

Metrics reported after generating fallback configuration and pushing it successfully:

```text
# HELP ingress_controller_processed_config_snapshot_cache_hit The number of times the controller hit the processed config snapshot cache and skipped generating a new one.
# TYPE ingress_controller_processed_config_snapshot_cache_hit counter
ingress_controller_processed_config_snapshot_cache_hit 1

# HELP ingress_controller_processed_config_snapshot_cache_miss The number of times the controller missed the processed config snapshot cache and had to generate a new one.
# TYPE ingress_controller_processed_config_snapshot_cache_miss counter
ingress_controller_processed_config_snapshot_cache_miss 1

# HELP ingress_controller_fallback_configuration_push_broken_resource_count The number of resources not accepted by Kong when attempting to push fallback configuration. `dataplane` describes the dataplane that was the target of the configuration push.
# TYPE ingress_controller_fallback_configuration_push_broken_resource_count gauge
ingress_controller_fallback_configuration_push_broken_resource_count{dataplane="http://127.0.0.1:52549"} 0

# HELP ingress_controller_fallback_cache_generating_duration_milliseconds How long it took to generate a fallback cache, in milliseconds. `success` describes whether the cache generation was successful (`true`) or not (`false`).
# TYPE ingress_controller_fallback_cache_generating_duration_milliseconds histogram
ingress_controller_fallback_cache_generating_duration_milliseconds_bucket{success="true",le="0.005"} 1
ingress_controller_fallback_cache_generating_duration_milliseconds_bucket{success="true",le="0.01"} 1
ingress_controller_fallback_cache_generating_duration_milliseconds_bucket{success="true",le="0.025"} 1
ingress_controller_fallback_cache_generating_duration_milliseconds_bucket{success="true",le="0.05"} 1
ingress_controller_fallback_cache_generating_duration_milliseconds_bucket{success="true",le="0.1"} 1
ingress_controller_fallback_cache_generating_duration_milliseconds_bucket{success="true",le="0.25"} 1
ingress_controller_fallback_cache_generating_duration_milliseconds_bucket{success="true",le="0.5"} 1
ingress_controller_fallback_cache_generating_duration_milliseconds_bucket{success="true",le="1"} 1
ingress_controller_fallback_cache_generating_duration_milliseconds_bucket{success="true",le="2.5"} 1
ingress_controller_fallback_cache_generating_duration_milliseconds_bucket{success="true",le="5"} 1
ingress_controller_fallback_cache_generating_duration_milliseconds_bucket{success="true",le="10"} 1
ingress_controller_fallback_cache_generating_duration_milliseconds_bucket{success="true",le="+Inf"} 1
ingress_controller_fallback_cache_generating_duration_milliseconds_sum{success="true"} 0
ingress_controller_fallback_cache_generating_duration_milliseconds_count{success="true"} 1

# HELP ingress_controller_fallback_configuration_push_count Count of successful/failed fallback configuration pushes to Kong. `dataplane` describes the dataplane that was the target of configuration push. `protocol` describes the configuration protocol (`db-less` or `deck`) in use. `success` describes whether there were unrecoverable errors (`false`) or not (`true`). `failure_reason` is populated in case of `success="false"` and describes the reason of failure (one of `conflict`, `network`, `other`).
# TYPE ingress_controller_fallback_configuration_push_count counter
ingress_controller_fallback_configuration_push_count{dataplane="http://127.0.0.1:52549",failure_reason="",protocol="db-less",success="true"} 1

# HELP ingress_controller_fallback_configuration_push_duration_milliseconds How long it took to push the fallback configuration to Kong, in milliseconds. `dataplane` describes the dataplane that was the target of configuration push. `protocol` describes the configuration protocol (`db-less` or `deck`) in use. `success` describes whether there were unrecoverable errors (`false`) or not (`true`).
# TYPE ingress_controller_fallback_configuration_push_duration_milliseconds histogram
ingress_controller_fallback_configuration_push_duration_milliseconds_bucket{dataplane="http://127.0.0.1:52549",protocol="db-less",success="true",le="0.005"} 0
ingress_controller_fallback_configuration_push_duration_milliseconds_bucket{dataplane="http://127.0.0.1:52549",protocol="db-less",success="true",le="0.01"} 0
ingress_controller_fallback_configuration_push_duration_milliseconds_bucket{dataplane="http://127.0.0.1:52549",protocol="db-less",success="true",le="0.025"} 0
ingress_controller_fallback_configuration_push_duration_milliseconds_bucket{dataplane="http://127.0.0.1:52549",protocol="db-less",success="true",le="0.05"} 0
ingress_controller_fallback_configuration_push_duration_milliseconds_bucket{dataplane="http://127.0.0.1:52549",protocol="db-less",success="true",le="0.1"} 0
ingress_controller_fallback_configuration_push_duration_milliseconds_bucket{dataplane="http://127.0.0.1:52549",protocol="db-less",success="true",le="0.25"} 0
ingress_controller_fallback_configuration_push_duration_milliseconds_bucket{dataplane="http://127.0.0.1:52549",protocol="db-less",success="true",le="0.5"} 0
ingress_controller_fallback_configuration_push_duration_milliseconds_bucket{dataplane="http://127.0.0.1:52549",protocol="db-less",success="true",le="1"} 0
ingress_controller_fallback_configuration_push_duration_milliseconds_bucket{dataplane="http://127.0.0.1:52549",protocol="db-less",success="true",le="2.5"} 1
ingress_controller_fallback_configuration_push_duration_milliseconds_bucket{dataplane="http://127.0.0.1:52549",protocol="db-less",success="true",le="5"} 1
ingress_controller_fallback_configuration_push_duration_milliseconds_bucket{dataplane="http://127.0.0.1:52549",protocol="db-less",success="true",le="10"} 1
ingress_controller_fallback_configuration_push_duration_milliseconds_bucket{dataplane="http://127.0.0.1:52549",protocol="db-less",success="true",le="+Inf"} 1
ingress_controller_fallback_configuration_push_duration_milliseconds_sum{dataplane="http://127.0.0.1:52549",protocol="db-less",success="true"} 2
ingress_controller_fallback_configuration_push_duration_milliseconds_count{dataplane="http://127.0.0.1:52549",protocol="db-less",success="true"} 1

# HELP ingress_controller_fallback_configuration_push_last The time of the last successful fallback configuration push. `dataplane` describes the dataplane that was the target of the configuration push.
# TYPE ingress_controller_fallback_configuration_push_last gauge
ingress_controller_fallback_configuration_push_last{dataplane="http://127.0.0.1:52549"} 1.7169937377764149e+09

# HELP ingress_controller_fallback_translation_broken_resource_count The number of resources that the controller cannot successfully translate to Kong configuration in fallback mode.
# TYPE ingress_controller_fallback_translation_broken_resource_count gauge
ingress_controller_fallback_translation_broken_resource_count 0

# HELP ingress_controller_fallback_translation_count Count of translations from Kubernetes state to Kong state in fallback mode. `success` describes whether there were unrecoverable errors (`false`) or not (`true`). Unrecoverable error in this case means KIC wasn't able to translate a Kubernetes object to Kong model.
# TYPE ingress_controller_fallback_translation_count counter
ingress_controller_fallback_translation_count{success="true"} 1
``` 

**Which issue this PR fixes**:

Closes #6082.

**Special notes for your reviewer**:

**PR Readiness Checklist**:

Complete these before marking the PR as `ready to review`:

- [x] the `CHANGELOG.md` release notes have been updated to reflect any significant (and particularly user-facing) changes introduced by this PR

### CODE REVIEW

File: test/envtest/metrics_envtest_test.go
Comment (pmalek) [inline test/envtest/metrics_envtest_test.go:1]: 👍

File: internal/metrics/prometheus.go [hunk 1]
```diff
@@ -223,6 +356,60 @@ func (c *CtrlFuncMetrics) RecordTranslationBrokenResources(count int) {
...
+}
+
+// RecordFallbackTranslationSuccess records a failed fallback configuration translation.
+func (c *CtrlFuncMetrics) RecordFallbackTranslationSuccess() {
+	c.FallbackTranslationCount.With(prometheus.Labels{
+		SuccessKey: SuccessTrue,
+	}).Inc()
+}
+
+func (c *CtrlFuncMetrics) RecordProcessedConfigSnapshotCacheHit() {
```
Comment (pmalek) [inline internal/metrics/prometheus.go:373]: godoc missing

File: internal/metrics/prometheus.go [hunk 2]
```diff
@@ -223,6 +356,60 @@ func (c *CtrlFuncMetrics) RecordTranslationBrokenResources(count int) {
...
+	c.ProcessedConfigSnapshotCacheHit.Inc()
+}
+
+func (c *CtrlFuncMetrics) RecordProcessedConfigSnapshotCacheMiss() {
```
Comment (pmalek) [inline internal/metrics/prometheus.go:377]: godoc missing

File: internal/metrics/prometheus.go [hunk 3]
```diff
@@ -223,6 +356,60 @@ func (c *CtrlFuncMetrics) RecordTranslationBrokenResources(count int) {
...
+
+// RecordFallbackPushFailure records a failed fallback configuration push.
+func (c *CtrlFuncMetrics) RecordFallbackPushFailure(p Protocol, duration time.Duration, dataplane string, brokenResourcesCount int, err error) {
+	dpOpt := withDataplane(dataplane)
+	c.recordFallbackPushDuration(p, duration, dpOpt, withFailure())
+	c.recordFallbackPushCount(p, dpOpt, withError(err))
+	c.recordFallbackPushBrokenResources(brokenResourcesCount, dpOpt)
+}
+
+func (c *CtrlFuncMetrics) RecordFallbackCacheGeneratingDuration(d time.Duration, err error) {
```
Comment (pmalek) [inline internal/metrics/prometheus.go:403]: godoc missing

File: internal/metrics/prometheus.go [hunk 4]
```diff
@@ -79,6 +87,19 @@ const (
...
 
+// Fallback config push metrics names.
+const (
+	MetricNameFallbackTranslationCount           = "ingress_controller_fallback_translation_count"
+	MetricNameFallbackTranslationBrokenResources = "ingress_controller_fallback_translation_broken_resource_count"
+	MetricNameFallbackConfigPushCount            = "ingress_controller_fallback_configuration_push_count"
+	MetricNameFallbackConfigPushSuccessTime      = "ingress_controller_fallback_configuration_push_last"
+	MetricNameFallbackConfigPushDuration         = "ingress_controller_fallback_configuration_push_duration_milliseconds"
+	MetricNameFallbackConfigPushBrokenResources  = "ingress_controller_fallback_configuration_push_broken_resource_count"
+	MetricNameFallbackCacheGeneratingDuration    = "ingress_controller_fallback_cache_generating_duration_milliseconds"
```
Comment (pmalek) [inline internal/metrics/prometheus.go:98]: Wouldn't this be clearer to call it: 

```suggestion
	MetricNameFallbackCacheGenerationDuration    = "ingress_controller_fallback_cache_generation_duration_milliseconds"
```
?

File: internal/dataplane/kong_client.go [hunk 5]
```diff
@@ -575,12 +578,14 @@ func (c *KongClient) tryRecoveringWithFallbackConfiguration(
...
 		c.logger.V(util.DebugLevel).Info("Translation failures occurred when building fallback data-plane configuration", "count", failuresCount)
+	} else {
+		c.prometheusMetrics.RecordFallbackTranslationBrokenResources(0)
+		c.prometheusMetrics.RecordFallbackTranslationSuccess()
 	}
 
-	// TODO: https://github.com/Kong/kubernetes-ingress-controller/issues/6082
-	// Expose Prometheus metrics for fallback configuration parsing result.
-
 	const isFallback = true
```
Comment (pmalek) [inline internal/dataplane/kong_client.go:589]: Did I miss it or was this discussed somewhere already? Do we want this `const` here?

Comment (czeslavo) [reply internal/dataplane/kong_client.go:589]: Yeah, it was discussed [here](https://github.com/Kong/kubernetes-ingress-controller/pull/6099#discussion_r1618451178).

**Review by pmalek — COMMENTED** [review]
This looks good apart from 1 issue which unfortunately I see for other metrics that we already expose as well.

We can merge this as is since what's proposed is consistent with what we already have but it might be a good topic for discussion.

**Review by pmalek — APPROVED** [review]

**Pull request merged**